test/interface.jl

```diff
@@ -55,4 +57,54 @@
     @test ps1 == ps2 
     @test ps1 != ps3  # comparison is order dependent
   end
-end
\ No newline at end of file
+end
+
+@testset "Grads" begin
+  @testset "algebra" begin
+    w = rand(2)
+    x1 = rand(2)
+    x2 = rand(2)
+    b = rand(2)
+
+    gs1 = gradient(() -> sum(w .* x1), Params([w])) 
+    gs2 = gradient(() -> sum(w .* x2), Params([w])) 
+  
+    @test .- gs1 isa Grads
+    @test gs1 .- gs2 isa Grads 
+    @test .+ gs1 isa Grads
+    @test gs1 .+ gs2 isa Grads 
+    @test 2 .* gs1 isa Grads 
+    @test gs1 .* 2 isa Grads 
+    @test gs1 ./ 2 isa Grads  
+    @test (gs1 .+ gs2)[w] ≈ gs1[w] .+ gs2[w] 
+
+    gs3 = gradient(() -> sum(w .* x1), Params([w, b])) # grad nothing with respect to b
+    gs4 = gradient(() -> sum(w .* x2 .+ b), Params([w, b])) 
+
+    @test .- gs3 isa Grads
+    @test gs3 .- gs4 isa Grads 
+    @test .+ gs3 isa Grads
+    @test gs3 .+ gs4 isa Grads 
+    @test 2 .* gs3 isa Grads 
+    @test gs3 .* 2 isa Grads 
+    @test gs3 ./ 2 isa Grads  
+    @test (gs3 .+ gs4)[w] ≈ gs3[w] .+ gs4[w]
+    @test (gs3 .+ gs4)[b] ≈ gs4[b] 
```

Comment:
Make sure in-place broadcasting works (not sure if it should currently).
```suggestion
    @test (gs3 .+ gs4)[b] ≈ gs4[b]
    gs5 .= gs3 + gs4
    @test gs5[w] ≈ gs3[w] .+ gs4[w]
    @test gs5[b] ≈ gs3[b]
```

Reply:
which method should I overload for broadcasted assignment?

Reply:
I think you will want to define `copyto!(dest::Grads, bc::Broadcasted)` (reference: https://docs.julialang.org/en/v1/manual/interfaces/#extending-in-place-broadcast-2)

Reply:
I didn't understand copyto! quite well, also we don't have broadcasted styles, so I implemented a solution with materialize!

Reply:
I think you want specifically `copyto!(dest::Grads, bc::Broadcasted{Nothing})` since we don't specialize on the style. I'm okay with `materialize!` since it doesn't require "indexing" the `Broadcasted` wrapper. We just want to make sure this is safe (I don't know, maybe @mcabbott)?

Reply:
https://docs.julialang.org/en/v1/manual/interfaces/#man-interfaces-broadcasting

Specifically says copyto is the interface the internal machinery uses, so that's the best one forward. Also yes, in place broadcasting can cause corner cases (and incorrect gradients). You could broadcast setindex directly.